File: jlox/src/craftinginterpreters/lox/Expr.java
package craftinginterpreters.lox;

import java.util.List;

abstract class Expr {
	static class Binary extends Expr {
		Binary ( Expr left, Token operator, Expr right) {
			this.Expr = Expr;
			this.operator = operator;
			this.right = right;
		}
		final  Expr left;
		final Token operator;
		final Expr right;
	}
	static class Grouping extends Expr {
		Grouping ( Expr expression) {
			this.Expr = Expr;
		}
		final  Expr expression;
	}
	static class Literal extends Expr {
		Literal ( Object value) {
			this.Object = Object;
		}
		final  Object value;
	}
	static class Unary extends Expr {
		Unary ( Token operator, Expr right) {
			this.Token = Token;
			this.right = right;
		}
		final  Token operator;
		final Expr right;
	}
}

File: glox/lox/expr.go
package lox

type Expr interface {
	// essentially a no-op function simply to distinguish Exprs from generic interface{}
	Expression() Expr
}

type ExprAssign struct {
	Name Token
	Value Expr
}

type ExprBinary struct {
	Left     Expr
	Operator Token
	Right    Expr
}

type ExprGrouping struct {
	Expr
}

type ExprLiteral struct {
	Value interface{}
}

type ExprUnary struct {
	Operator Token
	Right    Expr
}

type ExprVar struct {
	Name Token
}

func (e *ExprAssign) Expression() Expr { return e }

func (e *ExprBinary) Expression() Expr { return e }

func (e *ExprGrouping) Expression() Expr { return e }

func (e *ExprLiteral) Expression() Expr { return e }

func (e *ExprUnary) Expression() Expr { return e }

func (e *ExprVar) Expression() Expr { return e }
